code/scripting/ade.cpp

```diff
@@ -804,122 +804,144 @@ ade_indexer::ade_indexer(lua_CFunction func,
+			if (add_typeinfo)
+				sprintf(buf, "String [%s]", s);
+			else
+				buf = s;
+			break;
+
+		case LUA_TTABLE:
 		{
-			case LUA_TNIL:
-				strcat(stackdump, "NIL");
-				break;
-			case LUA_TNUMBER:
-				d = lua_tonumber(L, argnum);
-				sprintf(buf, "Number [%f]",d);
-				strcat(stackdump, buf);
-				break;
-			case LUA_TBOOLEAN:
-				b = lua_toboolean(L, argnum);
-				sprintf(buf, "Boolean [%d]",b);
-				strcat(stackdump, buf);
-				break;
-			case LUA_TSTRING:
-				s = (char *)lua_tostring(L, argnum);
-				sprintf(buf, "String [%s]",s);
-				strcat(stackdump, buf);
-				break;
-			case LUA_TTABLE:
+			if (lua_getmetatable(L, argnum))
 			{
-				if(lua_getmetatable(L, argnum))
-				{
-					lua_pushstring(L, "__adeid");
-					lua_rawget(L, -2);
-					if(lua_isnumber(L, -1))
-					{
-						sprintf(buf, "Table [%s]", getTableEntry((size_t)lua_tonumber(L, -1)).Name);
-						strcat(stackdump, buf);
-					}
-					else
-						strcat(stackdump, "non-default Table");
-					lua_pop(L, 2);	//metatable and nil/adeid
-				}
+				lua_pushstring(L, "__adeid");
+				lua_rawget(L, -2);
+				if (lua_isnumber(L, -1))
+					sprintf(buf, "Table [%s]", getTableEntry((size_t)lua_tonumber(L, -1)).Name);
 				else
-					strcat(stackdump, "Table w/ no metatable");
+					buf = "non-default Table";
+				lua_pop(L, 2);	//metatable and nil/adeid
+			}
+			else
+				buf = "Table w/ no metatable";
 
+			if (add_typeinfo)
+			{
 				//Maybe get first key?
-				char *firstkey = NULL;
+				const char* firstkey = nullptr;
 				lua_pushnil(L);
-				if(lua_next(L, argnum))
+				if (lua_next(L, argnum))
 				{
-					firstkey = (char *)lua_tostring(L, -2);
-					if(firstkey != NULL)
+					firstkey = lua_tostring(L, -2);
+					if (firstkey != nullptr)
 					{
-						strcat(stackdump, ", First key: [");
-						strcat(stackdump, firstkey);
-						strcat(stackdump, "]");
+						buf += ", First key: [";
+						buf += firstkey;
+						buf += "]";
 					}
 					lua_pop(L, 1);	//Key
 				}
 				lua_pop(L, 1);	//Nil
 			}
-				break;
-			case LUA_TFUNCTION:
-				strcat(stackdump, "Function");
-				{
-					char *upname = (char*)lua_getupvalue(L, argnum, ADE_FUNCNAME_UPVALUE_INDEX);
-					if(upname != NULL)
-					{
-						strcat(stackdump, " ");
-						strcat(stackdump, lua_tostring(L, -1));
-						strcat(stackdump, "()");
-						lua_pop(L, 1);
-					}
-				}
-				break;
-			case LUA_TUSERDATA:
-				if(lua_getmetatable(L, argnum))
-				{
-					lua_pushstring(L, "__adeid");
-					lua_rawget(L, -2);
-					if(lua_isnumber(L, -1))
-					{
-						sprintf(buf, "Userdata [%s]", getTableEntry((size_t)lua_tonumber(L, -1)).Name);
-					}
-					else
-						sprintf(buf, "non-default Userdata");
+			break;
+		}
 
-					lua_pop(L, 2);	//metatable and nil/adeid
-				}
+		case LUA_TFUNCTION:
+		{
+			buf = "Function";
+			auto upname = lua_getupvalue(L, argnum, ADE_FUNCNAME_UPVALUE_INDEX);
+			if (upname != nullptr)
+			{
+				buf += " ";
+				buf += lua_tostring(L, -1);
+				buf += "()";
```

Comment:
Personally I'd prefer this chunk and others like it to be using sprintf or similar rather than concatenating three times, but it's really a matter of taste. Looks good time me overall